src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -44,136 +48,191 @@ internal class MvcEndpointDataSource : EndpointDataSource
-            var endpoints = new List<Endpoint>();
-            StringBuilder patternStringBuilder = null;
+            get
+            {
+                Initialize();
+                Debug.Assert(_changeToken != null);
+                Debug.Assert(_endpoints != null);
+                return _endpoints;
+            }
+        }
 
-            foreach (var action in _actions.ActionDescriptors.Items)
+        public override IChangeToken GetChangeToken()
+        {
+            Initialize();
+            Debug.Assert(_changeToken != null);
+            Debug.Assert(_endpoints != null);
+            return _changeToken;
+        }
+
+        private void Initialize()
+        {
+            if (_endpoints == null)
             {
-                if (action.AttributeRouteInfo == null)
+                lock (_lock)
                 {
-                    // In traditional conventional routing setup, the routes defined by a user have a static order
-                    // defined by how they are added into the list. We would like to maintain the same order when building
-                    // up the endpoints too.
-                    //
-                    // Start with an order of '1' for conventional routes as attribute routes have a default order of '0'.
-                    // This is for scenarios dealing with migrating existing Router based code to Endpoint Routing world.
-                    var conventionalRouteOrder = 0;
-
-                    // Check each of the conventional patterns to see if the action would be reachable
-                    // If the action and pattern are compatible then create an endpoint with the
-                    // area/controller/action parameter parts replaced with literals
-                    //
-                    // e.g. {controller}/{action} with HomeController.Index and HomeController.Login
-                    // would result in endpoints:
-                    // - Home/Index
-                    // - Home/Login
-                    foreach (var endpointInfo in ConventionalEndpointInfos)
+                    if (_endpoints == null)
                     {
-                        if (MatchRouteValue(action, endpointInfo, "Area")
-                            && MatchRouteValue(action, endpointInfo, "Controller")
-                            && MatchRouteValue(action, endpointInfo, "Action"))
-                        {
-                            var newPathSegments = endpointInfo.ParsedPattern.PathSegments.ToList();
+                        UpdateEndpoints();
+                    }
+                }
+            }
+        }
+        
+        private void UpdateEndpoints()
+        {
+            lock (_lock)
+            {
+                var endpoints = new List<Endpoint>();
+                StringBuilder patternStringBuilder = null;
 
-                            for (var i = 0; i < newPathSegments.Count; i++)
+                foreach (var action in _actions.ActionDescriptors.Items)
```

Comment:
Everything inside of the lock is unchanged. The diff looks bad here, but I just tabbed it over.